File: bridges/nvda/addon/globalPlugins/nvdaMcpBridge/domain/entities/speech_buffer.py
# nvdaMcpBridge domain -- SpeechBuffer: indexed capture of what NVDA speaks.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
# ROLE: entity.
# FED BY: the SpeechSource port's implementation, calling append.
# READ BY: the speech command handlers.

from __future__ import annotations

from collections.abc import Callable
from typing import TYPE_CHECKING, Any

from .indexed_buffer import IndexedBuffer
from .speech_text import join_speech

if TYPE_CHECKING:
	from ..ports.clock import Clock
	from ..ports.continuous_read import ContinuousRead

# Neither capture mode has an exact finish signal, so speech counts as finished after this much quiet.
# NVDA 2026.1 finishes a keystroke's speech about 124 ms after the gesture, well inside this window.
SPEECH_FINISHED_SECONDS: float = 1.0

# Bounds how long a claimed continuous read may hold the settle open, so a wrong port answer cannot hang.
# NVDA 2026.1 with ibmeci measured 1.8 to 2.0 s between say-all chunks.
CONTINUOUS_READ_STALE_SECONDS: float = 6.0


class SpeechBuffer(IndexedBuffer):
	"""``exact_finish`` is left false in both capture modes, which use the elapsed-time heuristic."""

	def __init__(
		self,
		clock: Clock,
		*,
		exact_finish: bool = False,
		continuous_read: ContinuousRead | None = None,
	) -> None:
		super().__init__(clock)
		self.exact_finish: bool = exact_finish
		self._speaking: bool = False
		self._observer: Callable[[str], None] | None = None
		# None means no continuous read is ever in progress.
		self._continuous_read: ContinuousRead | None = continuous_read

	def _sentinel(self) -> Any:
		return [""]

	def _render(self, entry: Any) -> str:
		return join_speech(entry)

	def set_observer(self, observer: Callable[[str], None] | None) -> None:
		"""The observer is called outside the lock."""
		self._observer = observer

	def append(self, sequence: Any, log_position: int = 0) -> None:
		"""Called from NVDA's speech thread."""
		with self._lock:
			self._record(sequence, log_position)
			self._speaking = True
			text = join_speech(sequence)
		if self._observer is not None and text:
			self._observer(text)

	def notify_finished(self) -> None:
		with self._lock:
			self._speaking = False

	def index_of(self, text: str, after_index: int | None = None) -> int:
		"""``after_index`` is an inclusive left edge, clamped to 0. Returns ``-1`` when not found."""
		first = 0 if after_index is None else max(0, after_index)
		with self._lock:
			for offset, entry in enumerate(self._entries[first:]):
				if text in self._render(entry):
					return first + offset
		return -1

	def wait_for(self, text: str, after_index: int | None, timeout: float) -> tuple[bool, int, str]:
		"""Returns ``(found, index, text)``; on a miss, ``index`` is the current :meth:`next_index`."""
		found_index = -1

		def _seen() -> bool:
			nonlocal found_index
			found_index = self.index_of(text, after_index)
			return found_index >= 0

		if self._wait(_seen, timeout):
			with self._lock:
				return True, found_index, self._render(self._entries[found_index])
		return False, self.next_index(), ""

	def collect_since(self, index: int, grace: float) -> tuple[list[tuple[str, int, int, float]], int, int]:
		"""Return :meth:`entries_since`'s triple once anything arrives or ``grace`` elapses.

		Returns early, so an utterance still in flight is left for the next read. An empty result means
		nothing had arrived by then, not that nothing will.
		"""

		def _arrived() -> bool:
			entries, _from_index, _to_index = self.entries_since(index)
			return bool(entries)

		self._wait(_arrived, grace)
		return self.entries_since(index)

	def wait_to_finish(self, timeout: float) -> bool:
		return self._wait(self._has_finished, timeout)

	def _has_finished(self) -> bool:
		# Outside the lock, so a port call never holds the mutex append() takes on NVDA's thread.
		if self._continuous_read is not None and self._continuous_read.in_progress():
			with self._lock:
				quiet_for = self._clock.monotonic() - self._last_time
			# Believed, but not forever: see CONTINUOUS_READ_STALE_SECONDS.
			if quiet_for <= CONTINUOUS_READ_STALE_SECONDS:
				return False
		with self._lock:
			if self.exact_finish:
				return not self._speaking
			return (self._clock.monotonic() - self._last_time) > SPEECH_FINISHED_SECONDS

File: bridges/voiceover/Sources/VoiceOverBridgeDomain/Entities/SpeechBuffer.swift
// ROLE: entity -- the indexed capture of what the reader said.
// FED BY: the SpeechSource implementation, on the source's own thread.
// READ BY: the five speech handlers, on the session thread.
// The index is this array's position, never the extension's own sequence number, which restarts when the system relaunches the extension.
// Index 0 is an empty sentinel, so the first capture lands at 1 and `getLastSpeech` on an untouched session answers with an empty string.
// Every read clamps: a stale bookmark answers with the sentinel or an empty range, never an error.

import Foundation

public let speechPollInterval: Double = 0.03

/// There is no exact finished signal (silent mode produces no audio), so this waits for the buffer to stop growing.
public let speechFinishedSeconds: Double = 1.0

public final class SpeechBuffer {
	private let clock: any Clock
	/// Recursive because `waitFor` reads through `indexOf` while holding the lock.
	private let lock = NSRecursiveLock()

	/// Append-only and unbounded within a session.
	private var entries: [CapturedUtterance]

	/// Monotonic, so the elapsed-time heuristic survives a clock correction.
	private var lastAppendedAt: Double

	private var observer: ((String) -> Void)?

	public init(clock: any Clock) {
		self.clock = clock
		self.entries = [CapturedUtterance(text: "")]
		self.lastAppendedAt = clock.monotonic()
	}

	/// Fired outside the lock: the transcript writes to a file, and IO under the mutex would stall the capture thread.
	public func setObserver(_ observer: ((String) -> Void)?) {
		lock.lock()
		defer { lock.unlock() }
		self.observer = observer
	}

	/// Called from the speech source's thread.
	public func append(_ utterance: CapturedUtterance) {
		let notify: ((String) -> Void)?
		lock.lock()
		entries.append(utterance)
		lastAppendedAt = clock.monotonic()
		notify = observer
		lock.unlock()
		if !utterance.text.isEmpty {
			notify?(utterance.text)
		}
	}

	public func lastIndex() -> Int {
		lock.lock()
		defer { lock.unlock() }
		return entries.count - 1
	}

	public var isEmpty: Bool {
		lock.lock()
		defer { lock.unlock() }
		return entries.count <= 1
	}

	public func nextIndex() -> Int {
		lock.lock()
		defer { lock.unlock() }
		return entries.count
	}

	public func last() -> (utterance: CapturedUtterance, index: Int) {
		lock.lock()
		defer { lock.unlock() }
		return (entries[entries.count - 1], entries.count - 1)
	}

	public func entry(at index: Int) -> CapturedUtterance {
		lock.lock()
		defer { lock.unlock() }
		guard index >= 0, index < entries.count else { return entries[0] }
		return entries[index]
	}

	/// Returns the half-open range read, not the span returned: empty entries are skipped, so each carries its own index.
	public func entriesSince(_ index: Int) -> (
		entries: [(utterance: CapturedUtterance, index: Int)], fromIndex: Int, toIndex: Int
	) {
		lock.lock()
		defer { lock.unlock() }
		let from = max(0, index)
		let to = entries.count
		var found: [(utterance: CapturedUtterance, index: Int)] = []
		for position in from..<max(from, to) where !entries[position].text.isEmpty {
			found.append((entries[position], position))
		}
		return (found, from, to)
	}

	/// An empty result means nothing had arrived by the deadline, not that nothing will; returns as soon as anything with words arrives.
	public func collectSince(_ index: Int, grace: Double) -> (
		entries: [(utterance: CapturedUtterance, index: Int)], fromIndex: Int, toIndex: Int
	) {
		_ = wait(timeout: grace) { !self.entriesSince(index).entries.isEmpty }
		return entriesSince(index)
	}

	/// `afterIndex` is inclusive: an exclusive reading drops the first utterance an action caused.
	public func indexOf(_ text: String, afterIndex: Int? = nil) -> Int? {
		lock.lock()
		defer { lock.unlock() }
		let first = max(0, afterIndex ?? 0)
		guard first < entries.count else { return nil }
		for position in first..<entries.count where entries[position].text.contains(text) {
			return position
		}
		return nil
	}

	/// On a miss the index is a fresh bookmark and the utterance is empty.
	public func waitFor(_ text: String, afterIndex: Int?, timeout: Double) -> (
		found: Bool, index: Int, utterance: CapturedUtterance
	) {
		var hit: Int?
		let seen = wait(timeout: timeout) {
			hit = self.indexOf(text, afterIndex: afterIndex)
			return hit != nil
		}
		if seen, let hit {
			return (true, hit, entry(at: hit))
		}
		return (false, nextIndex(), CapturedUtterance(text: ""))
	}

	public func waitToFinish(timeout: Double) -> Bool {
		wait(timeout: timeout) { self.hasFinished() }
	}

	private func hasFinished() -> Bool {
		lock.lock()
		defer { lock.unlock() }
		return (clock.monotonic() - lastAppendedAt) > speechFinishedSeconds
	}

	private func wait(timeout: Double, _ predicate: () -> Bool) -> Bool {
		let deadline = clock.monotonic() + max(0, timeout)
		while true {
			if predicate() { return true }
			if clock.monotonic() >= deadline { return false }
			clock.sleep(speechPollInterval)
		}
	}
}
